[cocoa] Define and add new Error to surface failed main resource loads from privacy proxy

sysrqb opened the pull request:
#### 6d48ee17b3268ad1a688734f135400bbfe98bd73
<pre>
[cocoa] Define and add new Error to surface failed main resource loads from privacy proxy
<a href="https://bugs.webkit.org/show_bug.cgi?id=248893">https://bugs.webkit.org/show_bug.cgi?id=248893</a>
rdar://problem/103082081

Reviewed by Wenson Hsieh.

Expose the provided &quot;privacy stance&quot; in the NSError as a userinfo dictionary
entry;

Tested manually.

* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(privacyStanceToString):
(-[WKNetworkSessionDelegate URLSession:task:didCompleteWithError:]):

Canonical link: <a href="https://commits.webkit.org/257664@main">https://commits.webkit.org/257664@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9319eb146898f0063d4416752f472b91f95db0f4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99601 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8776 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32693 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108961 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9378 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86077 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92075 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106877 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105375 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90593 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [💥 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34038 "An unexpected error occured. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Updated gtk dependencies; Pull request contains relevant changes; Skipped layout-tests; layout-tests (exception)") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88892 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21950 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77303 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2611 "Build is being retried. Recent messages:Cleaned up git repository; 'python3 Tools/Scripts/git-webkit ...'; Skipping applying patch since patch_id isn't provided; Checked out pull request; compiling") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23464 "Passed tests") | | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2547 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8690 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42937 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5274 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4409 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
```diff
@@ -788,6 +806,10 @@ - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didComp
         newUserInfo[@"networkTaskDescription"] = [task description];
+#if HAVE(NETWORK_CONNECTION_PRIVACY_STANCE)
+        if (auto* networkDataTask = [self existingTask:task])
+            newUserInfo[@"networkTaskMetricsPrivacyStance"] = toPrivacyStanceToString(networkDataTask->networkLoadMetrics().privacyStance);
```
```suggestion
            newUserInfo[@"networkTaskMetricsPrivacyStance"] = privacyStanceToString(networkDataTask->networkLoadMetrics().privacyStance);
```

sysrqb replied:
Yeah, good suggestion.

whsieh commented on Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
```diff
@@ -163,6 +163,24 @@ static NSURLSessionAuthChallengeDisposition toNSURLSessionAuthChallengeDispositi
     return WebCore::PrivacyStance::Unknown;
 }
+
+static NSString* toPrivacyStanceToString(WebCore::PrivacyStance stance)
```
```suggestion
static NSString *privacyStanceToString(WebCore::PrivacyStance stance)
```

sysrqb committed:
[cocoa] Define and add new Error to surface failed main resource loads from privacy proxy
https://bugs.webkit.org/show_bug.cgi?id=248893
rdar://problem/103082081

Reviewed by Wenson Hsieh.

Expose the provided "privacy stance" in the NSError as a userinfo dictionary
entry;

Tested manually.

* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(privacyStanceToString):
(-[WKNetworkSessionDelegate URLSession:task:didCompleteWithError:]):

Canonical link: https://commits.webkit.org/257664@main
